File: go/parsers/publish.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Publish struct {
	gen     *gen.Publish
	payload []byte
}

func NewPublishFields(g *gen.Publish, payload []byte) messages.PublishFields {
	return &Publish{gen: g, payload: payload}
}

func (p *Publish) RequestID() uint64 {
	return p.gen.RequestID()
}

func (p *Publish) Topic() string {
	topic, _ := p.gen.Topic()
	return topic
}

func (p *Publish) Options() map[string]any {
	return map[string]any{}
}

func (p *Publish) Args() []any {
	return nil
}

func (p *Publish) KwArgs() map[string]any {
	return nil
}

func (p *Publish) PayloadIsBinary() bool {
	return true
}

func (p *Publish) Payload() []byte {
	return nil
}

func (p *Publish) PayloadSerializer() uint64 {
	return 0
}

func PublishToCapnproto(m *messages.Publish) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	publish, err := gen.NewRootPublish(seg)
	if err != nil {
		return nil, err
	}

	publish.SetRequestID(m.RequestID())
	if err := publish.SetTopic(m.Topic()); err != nil {
		return nil, err
	}

	data, err := msg.Marshal()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypePublish, data), nil
}

func CapnprotoToPublish(data, payload []byte) (*messages.Publish, error) {
	msg, err := capnp.Unmarshal(data)
	if err != nil {
		return nil, err
	}

	publish, err := gen.ReadRootPublish(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewPublishWithFields(NewPublishFields(&publish, payload)), nil
}

File: python/wamp_msgs_capnp/parsers/publish.py
import os
from typing import Any, Optional
from pathlib import Path

import capnp
from wampproto.messages.publish import Publish, IPublishFields
from wampproto.serializers.payload import serialize_payload

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
module_file = os.path.join(root_dir, "schemas", "publish.capnp")
publish_capnp = capnp.load(str(module_file))


class PublishFields(IPublishFields):
    def __init__(self, gen, payload: bytes):
        self._gen = gen
        self._ex = helpers.PayloadExpander(payload, gen.payloadSerializerID)

    @property
    def request_id(self) -> int:
        return self._gen.requestID

    @property
    def topic(self) -> str:
        return self._gen.topic

    @property
    def options(self) -> Optional[dict[str, Any]]:
        details = {}

        if not self._gen.excludeMe:
            details["exclude_me"] = False

        return details

    @property
    def args(self) -> list[Any]:
        return self._ex.args

    @property
    def kwargs(self) -> dict[str, Any]:
        return self._ex.kwargs

    @property
    def payload_is_binary(self) -> bool:
        return True

    @property
    def payload(self) -> bytes:
        return self._ex.payload

    @property
    def payload_serializer_id(self) -> int:
        return self._gen.payloadSerializerID


def publish_to_capnproto(p: Publish) -> bytes:
    publish = publish_capnp.Publish.new_message()

    publish.requestID = p.request_id
    publish.topic = p.topic

    payload_serializer = helpers.select_payload_serializer(p.options)
    publish.payloadSerializerID = payload_serializer

    if p.payload_is_binary():
        payload = p.payload
    else:
        payload = serialize_payload(payload_serializer, p.args, p.kwargs)

    packed_data = publish.to_bytes_packed()

    return helpers.prepend_header(Publish.TYPE, packed_data, payload)


def capnproto_to_publish(data: bytes) -> Publish:
    message_data, payload_data = helpers.extract_message(data)
    publish_obj = publish_capnp.Publish.from_bytes_packed(message_data)

    return Publish(PublishFields(publish_obj, payload_data))
